File: Leetcode/Python/0581_Shortest_Unsorted_Continuous_Subarray.py
class Solution(object):
    def findUnsortedSubarray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        smallest = nums[0]
        largest = nums[-1]
        left = 0
        right = 0
        # the most chaos, the whole list need to be changed
        if smallest != min(nums) and largest != max(nums):
            return len(nums)
        if smallest != min(nums):
            return nums.index(min(nums)) + 1
        if largest != max(nums):
            return len(nums) - nums.index(max(nums))
        if nums == sorted(nums):
            return 0
        for i, x in enumerate(nums):
            if x < smallest:
                left = i - 1
                break
            smallest = x
        nums = nums[::-1]
        for i, x in enumerate(nums):
            if x > largest:
                right = len(nums) - i
                break
            largest = x
        return right - left + 1

File: Leetcode/Java/0581_Shortest_Unsorted_Continuous_Subarray.java
class Solution {
    public int findUnsortedSubarray_sortedArray(int[] nums) {
        int[] numsCopy = nums.clone();
        Arrays.sort(numsCopy);
        int start = nums.length;
        int end = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != numsCopy[i]) {
                start = Math.min(i, start);
                end = Math.max(i, end);
            }
        }
        return (end - start >= 0) ? end - start + 1 : 0;
    }

    public int findUnsortedSubarray_stack(int[] nums) {
        if (nums.length <= 1) {
            return 0;
        }
        Stack<Integer> stack = new Stack<>();
        int left = nums.length - 1;
        int right = 0;
        for (int i = 0; i < nums.length; i++) {
            if (stack.isEmpty() || nums[i] >= nums[stack.peek()]) {
                stack.push(i);
            } else {
                left = Math.min(left, stack.pop());
                i--;
            }
        }
        stack.clear();
        for (int i = nums.length - 1; i >= 0; i--) {
            if (stack.isEmpty() || nums[i] <= nums[stack.peek()]) {
                stack.push(i);
            } else {
                right = Math.max(right, stack.pop());
                i++;
            }
        }
        return right - left > 0 ? right - left + 1 : 0;
    }
}